rfcs/0006-transaction.md

```diff
@@ -0,0 +1,301 @@
+4. The conflict check verifies that: the transaction's read and write keys do not conflict with other transactions' write keys
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
+
+(note: at the time of this RFC, the Snapshot API and WriteBatch API are still not implemented, the transaction feature should be implemented after these APIs are implemented.)
+
+The write operations like `put()` in the transaction should be buffered in the transaction state, and only be written to the underlying storage when `commit()` is called.
+
+```rust
+let object_store: Arc<dyn ObjectStore> = Arc::new(InMemory::new());
+let options = DbOptions::default();
+let kv_store = Db::open_with_opts(
+    Path::from("/tmp/test_kv_store"),
+    options,
+    object_store,
+)
+.await
+.unwrap();
+
+// start transaction and write some data
+let txn = kv_store.transaction();
+let val1 = txn.get("key1").await;
+
+// could batch the write operations
+txn.put("key2", val1);
+txn.put("key3", "val3");
+
+// should see the new value of key2
+assert_eq!(txn.get("key2").await, val1);
+
+// commit, might fail on conflict, when it fails, the transaction should be rolled back
+txn.commit().await?;
+```
+
+### Conflict Checking
+
+The approach we choose on conflict checking is expected to be more similar to Badger's global `Oracle` approach, since we hope to provide SSI in the transaction feature, and it's less prone to unnecessarily abort the transaction with limited MemTable size.
+
+Below is the data structure of the global `Oracle` and the transaction state:
+
+```rust
+struct Oracle {
+    next_seq: AtomicU64,
+    recent_committed_txns: Deque<Arc<TransactionState>>,
+}
+
+type KeyFingerPrint = u32;
+
+struct TransactionState {
+    started_seq: u64,
+    write_keys: HashSet<KingFingerPrint>,
+    read_keys: HashSet<KeyFingerPrint>,
```

Comment:
I don't think this is sufficient to guarantee serializability (maybe I'm misunderstanding SSI) because it doesn't handle the case where new data is inserted that conflicts with a range scan. For example, suppose Txn1 reads data in the range "aaa", to "bbb" and gets rows with keys "aaa" and "aab". Meanwhile Txn2 inserts "aac", which conflicts with Txn1 because it adds a row that should have been returned by Txn1's read. Txn1 would still be able to commit. To mitigate this, shouldn't we also retain ranges of keys that each txn reads and check the writes against those ranges?

Reply:
this is a good point, this approach reflect what badger does, but it seems badger did not take the range scan into consideration. 

having ranges checks here is not expensive at the first glance, let me think about it and update the doc.

Reply:
One drawback is that we have to store entire keys for writes, rather than fingerprints. That said, I'd expect there to be relatively few keys written, and we have to store them anyway to buffer them for write batches - this just means we have to hold on to them a little longer to check conflicts.

Reply:
updated in 1d74fee

Reply:
imho the keys & values in the `WriteBatch` is better to seperated in different `Bytes`, the `value`s are likely to be bigger than the `key`s, only referencing the `key`s in `WriteBatch` is better for saving the memory usage.